### PR TITLE
New deploy staker

### PR DESCRIPTION
**Pull request by robertg93**

### CODE REVIEW

**Review by wojciech-cichocki — CHANGES_REQUESTED** [review]

File: sdk/src/math.ts [hunk 1]
```diff
@@ -38,6 +38,9 @@ export const isInitialized = (tickmap: Tickmap, index: number, tickSpacing: numb
 
   return value !== 0
 }
+export const priceLog = (val): number => {
```
[the remark this answers is not in the document]
Comment (robertg93) [reply sdk/src/math.ts:41]: done

File: staker-sdk/src/utils.ts [hunk 2]
```diff
@@ -30,7 +25,7 @@ export const toDecimal = (x: number, decimals: number = 0): Decimal => {
   return { v: DENOMINATOR.muln(x).div(new BN(10).pow(new BN(decimals))) }
 }
 
-export const STAKER_SEED = 'staker'
+export const STAKER_SEED = Buffer.from('staker')
```
[the remark this answers is not in the document]
Comment (robertg93) [reply staker-sdk/src/utils.ts:28]: done

File: sdk/src/utils.ts [hunk 3]
```diff
@@ -750,3 +756,176 @@ export const getMinTick = (tickSpacing: number) => {
...
+    liquidity.v.toNumber() /
+    LIQUIDITY_DENOMINATOR.toNumber()
+  )
+}
+export const poolAPY = (dailyFactorPool: number): number => {
+  return (Math.pow(dailyFactorPool + 1, 365) - 1) * 100
+}
+
+export const getRangeFromPrices = (
+  priceMinA: number,
```
[the remark this answers is not in the document]
Comment (robertg93) [reply sdk/src/utils.ts:773]: done

File: sdk/src/utils.ts [hunk 4]
```diff
@@ -750,3 +756,176 @@ export const getMinTick = (tickSpacing: number) => {
...
+  liquidity: BN
+  index: number
+}
+
+export interface LiquidityRange {
+  tickLower: number
+  tickUpper: number
+}
+
+export const TICKS = [
```
[the remark this answers is not in the document]
Comment (robertg93) [reply sdk/src/utils.ts:890]: done

File: sdk/src/utils.ts [hunk 5]
```diff
@@ -750,3 +756,176 @@ export const getMinTick = (tickSpacing: number) => {
...
+export const dailyFactorPool = (volume: Decimal, liquidity: Decimal, feeTier: FeeTier): number => {
+  const fee: number = (feeTier.fee.toNumber() / FEE_TIER_DENOMINATOR) * (1 - PROTOCOL_FEE)
+  return (
+    (volume.v.toNumber() * fee * LIQUIDITY_DENOMINATOR.toNumber()) /
+    liquidity.v.toNumber() /
+    LIQUIDITY_DENOMINATOR.toNumber()
+  )
+}
+export const poolAPY = (dailyFactorPool: number): number => {
+  return (Math.pow(dailyFactorPool + 1, 365) - 1) * 100
```
[the remark this answers is not in the document]
Comment (robertg93) [reply sdk/src/utils.ts:769]: just for frontend usage

File: programs/invariant/src/structs/tick.rs [hunk 6]
```diff
@@ -15,7 +15,6 @@ pub struct Tick {
     pub fee_growth_outside_x: FeeGrowth,
     pub fee_growth_outside_y: FeeGrowth,
     pub seconds_per_liquidity_outside: FixedPoint,
-    pub seconds_outside: u64,
```
[the remark this answers is not in the document]
Comment (robertg93) [reply programs/invariant/src/structs/tick.rs:18]: done

File: programs/invariant/src/util.rs [hunk 7]
```diff
@@ -92,8 +92,6 @@ pub fn cross_tick(tick: &mut RefMut<Tick>, pool: &mut Pool) -> Result<()> {
         .unchecked_sub(tick.fee_growth_outside_y);
 
     let current_timestamp = get_current_timestamp();
-    let seconds_passed: u64 = current_timestamp.checked_sub(pool.start_timestamp).unwrap();
```
[the remark this answers is not in the document]
Comment (robertg93) [reply programs/invariant/src/util.rs:95]: done

File: sdk/package-lock.json [hunk 8]
```diff
@@ -1,2980 +0,0 @@
-{
```
[the remark this answers is not in the document]
Comment (robertg93) [reply sdk/package-lock.json:1]: done

File: sdk/src/utils.ts [hunk 9]
```diff
@@ -750,3 +770,158 @@ export const getMinTick = (tickSpacing: number) => {
...
+  let tickLower: number = null
+  let tickUpper: number = null
+  let tickLowerSaved = false
+  let lastIndex = 0
+
+  for (let i = 0; i < tickArrayPrevious.length - 1; i++) {
+    let previousTick = tickArrayPrevious[i]
+    let currentTick = tickArrayCurrent[i]
+
+    if (previousTick.feeGrowth != currentTick.feeGrowth) {
```
Comment (wojciech-cichocki) [inline sdk/src/utils.ts:888]: use `equals` function to compare `BN` types

File: sdk/src/utils.ts [hunk 10]
```diff
@@ -750,3 +770,158 @@ export const getMinTick = (tickSpacing: number) => {
...
+  tickArrayCurrent: ParsedTick[]
+) => {
+  let tickLower: number = null
+  let tickUpper: number = null
+  let tickLowerSaved = false
+  let lastIndex = 0
+
+  for (let i = 0; i < tickArrayPrevious.length - 1; i++) {
+    let previousTick = tickArrayPrevious[i]
+    let currentTick = tickArrayCurrent[i]
```
Comment (wojciech-cichocki) [inline sdk/src/utils.ts:886]: this for sure should't work. Find current tick state based on the same index. If current tick state not exist - don not mark this tick as changed

File: sdk/src/utils.ts [hunk 11]
```diff
@@ -750,3 +770,158 @@ export const getMinTick = (tickSpacing: number) => {
...
+        tickLowerSaved = true
+        lastIndex = i
+      }
+      tickUpper = currentTick.index
+    }
+  }
+  if (tickLower == tickUpper) {
+    tickUpper = tickArrayPrevious[lastIndex + 1].index
+  }
+  return {
```
Comment (wojciech-cichocki) [inline sdk/src/utils.ts:900]: What this function returns if new snapshot contain no positions?

Comment (robertg93) [reply sdk/src/utils.ts:900]: there is a check for this in function which calls this one

**Review by wojciech-cichocki — APPROVED** [review]

**Pull request merged**